databuilder/query_engines/sqlite.py

```diff
@@ -213,13 +220,37 @@ def aggregate_by_patient(self, source_node, aggregation_func):
         aggregated_table = self.reify_query(query)
         return aggregated_table.c.value
 
-    def aggregate_by_patient_non_nullable(self, node, aggregation_func, empty_value):
-        # These aggregation functions don't take an argument as they operate over the
-        # entire query, so we wrap them in a lambda which swallows the argument
-        value = self.aggregate_by_patient(node, lambda _: aggregation_func)
-        # These aggregations are also guaranteed not to be null-valued, even when
-        # combined with data involving patient_ids for which they're not defined
-        return sqlalchemy.func.coalesce(value, empty_value)
+    def aggregate_frame_by_patient(
+        self, source_node, aggregation_func, empty_value, single_row_value
+    ):
+        # Frame aggregations can operate on both one- and many-rows-per-patient frames
+        # and these cases require different handling.
+        if has_many_rows_per_patient(source_node):
+            # For many-rows-per-patient frames we use the series aggregation code but
+            # wrapped in some extra logic: these aggregation functions don't take an
+            # argument as they operate over the entire query, so we wrap them in a
+            # lambda which swallows the argument
+            value = self.aggregate_series_by_patient(
+                source_node, lambda _: aggregation_func
+            )
+            # These aggregations are also guaranteed not to be null-valued, even when
```

Comment:
I know you haven't changed this comment, but I think I remember this bringing me to a halt before. The way it's phrased means it's ambiguous whether "these aggregations" refers to `aggregation_func` (it doesn't) or the aggregation that this method is being invoked to implement (it does).

Reply:
Good point. I've changed to this:
```
            # We also want to guarantee that our frame level aggregations are never
            # null-valued, even when combined with data involving patient_ids for which
            # they're not defined
```

Reply:
Thanks